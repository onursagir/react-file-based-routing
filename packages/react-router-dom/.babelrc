{
  "extends": "../../.babelrc",
  "ignore": ["src/__tests__/**/*", "src/__stubs__/**/*", "src/**/macro.d.ts"]
}
