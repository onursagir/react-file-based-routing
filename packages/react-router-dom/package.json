{
  "name": "@react-file-based-routing/react-router-dom",
  "version": "1.0.0",
  "description": "",
  "main": "dist",
  "scripts": {
    "dev": "npm run build -- --watch",
    "lint": "eslint --fix src/**",
    "format": "prettier --write 'src/**/*.{js,jsx,ts,tsx}'",
    "test": "jest --config ../../jest.config.ts",
    "test:watch": "npm run test -- --watch",
    "build": "rimraf dist && babel src --out-dir dist --extensions .js,.ts && cpy 'src/macro.d.ts' 'dist' --flat"
  },
  "keywords": [
    "react",
    "file",
    "based",
    "routing",
    "react-router-dom"
  ],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "fast-glob": "^3.2.11",
    "remove": "^0.1.5"
  },
  "peerDependencies": {
    "react-router-dom": "^6.3.0"
  },
  "devDependencies": {
    "cpy-cli": "^4.1.0"
  }
}
