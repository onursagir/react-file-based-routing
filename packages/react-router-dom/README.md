## Usage

> ⚠️ Requires [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) to work.

Simply specify the directory containing your routes using the macro like so `<FileBasedRoutes path="./pages" />`. Do note that the actual name of the folder will not be included in the route path. For example the file `pages/user/profile.tsx` will result in the route path `user/profile`. Regarding index routes, route-nesting and dynamic route segments the aim is to follow the [NextJs Routing](https://nextjs.org/docs/routing/introduction) as closely as possible. See below for further explanation

### Index routes

- `pages/index.js` → `/`
- `pages/blog/index.js` → `blog`

### Nested routes

- `pages/blog/first-post.js` → `/blog/first-post`
- `pages/dashboard/settings/username.js` → `/dashboard/settings/username`

### Dynamic route segments

- `pages/blog/[slug].js` → `/blog/:slug`
- `pages/[username]/settings.js` → `/:username/settings`
- `pages/post/[...all].js` → `/post/*`

## API

| Property  | Description                                                                           | Type    | Default |
| --------- | ------------------------------------------------------------------------------------- | ------- | ------- |
| path      | Relative path to the folder you want to have the routes based out from                | string  | -       |
| logRoutes | Will output the routes that have been made along with their components to the console | boolean | false   |

---

## Examples

```TypeScript
import React from "react";
import ReactDOM from "react-dom/client";
import { BrowserRouter, Routes } from "react-router-dom";
import FileBasedRoutes from "@react-file-based-routing/react-router-dom/macro";

ReactDOM.createRoot(document.getElementById("root")!).render(
  <React.StrictMode>
    <BrowserRouter>
      <Routes>
        <FileBasedRoutes path="./pages" />
        <FileBasedRoutes path="./some-other-folder" logRoutes />
      </Routes>
    </BrowserRouter>
  </React.StrictMode>
);

```
