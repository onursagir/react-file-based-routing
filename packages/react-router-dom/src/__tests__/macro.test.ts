import pluginTester from 'babel-plugin-tester';
import macrosPlugin from 'babel-plugin-macros';

jest.mock('../get-routes-and-sources.ts', () => ({
  getRoutesAndSources: jest.fn().mockReturnValue([
    {
      routerPath: 'foo',
      importSpecifierName: 'foo',
      targetRelativeToSource: './dir/foo',
    },
    {
      routerPath: 'bar/baz',
      importSpecifierName: 'bar_baz',
      targetRelativeToSource: './dir/bar/baz',
    },
  ]),
}));

describe('macro.test.ts', () => {
  pluginTester({
    plugin: macrosPlugin,
    snapshot: true,
    babelOptions: { filename: __filename, presets: ['@babel/preset-react'] },
    tests: [
      `
        import { BrowserRouter, Routes } from 'react-router-dom';
        import FileBasedRouting from '../macro';

        const App = () => (
          <BrowserRouter>
              <FileBasedRouting path="./some/folder" />
          </BrowserRouter>
        );
      `,
      `
        import FileBasedRouting from '../macro';
        import { BrowserRouter, Routes } from 'react-router-dom';

        const App = () => React.createElement(BrowserRouter, null, React.createElement(FileBasedRouting, {
          path: "./some/folder"
        }));
      `,
    ],
  });
});
