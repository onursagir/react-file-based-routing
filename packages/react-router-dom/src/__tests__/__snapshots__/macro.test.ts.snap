// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro.test.ts unknown plugin 1. unknown plugin: 1. unknown plugin 1`] = `

import { BrowserRouter, Routes } from 'react-router-dom';
import FileBasedRouting from '../macro';

const App = () => (
  <BrowserRouter>
      <FileBasedRouting path="./some/folder" />
  </BrowserRouter>
);

      ↓ ↓ ↓ ↓ ↓ ↓

import foo from './dir/foo';
import bar_baz from './dir/bar/baz';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

const App = () =>
  /*#__PURE__*/ React.createElement(BrowserRouter, null, [
    /*#__PURE__*/ React.createElement(Route, {
      key: '0',
      path: 'foo',
      element: /*#__PURE__*/ React.createElement('foo', null),
    }),
    /*#__PURE__*/ React.createElement(Route, {
      key: '1',
      path: 'bar/baz',
      element: /*#__PURE__*/ React.createElement('bar_baz', null),
    }),
  ]);


`;

exports[`macro.test.ts unknown plugin 2. unknown plugin: 2. unknown plugin 1`] = `

import FileBasedRouting from '../macro';
import { BrowserRouter, Routes } from 'react-router-dom';

const App = () => React.createElement(BrowserRouter, null, React.createElement(FileBasedRouting, {
  path: "./some/folder"
}));

      ↓ ↓ ↓ ↓ ↓ ↓

import foo from './dir/foo';
import bar_baz from './dir/bar/baz';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

const App = () =>
  React.createElement(BrowserRouter, null, [
    /*#__PURE__*/ React.createElement(Route, {
      key: '0',
      path: 'foo',
      element: /*#__PURE__*/ React.createElement('foo', null),
    }),
    /*#__PURE__*/ React.createElement(Route, {
      key: '1',
      path: 'bar/baz',
      element: /*#__PURE__*/ React.createElement('bar_baz', null),
    }),
  ]);


`;
